### ISSUE TITLE
Macro error with futures::try_join!

### ISSUE THREAD

**Issue by lnicola** [rust-lang/rust-analyzer#6835]
Same as @jeromegn , I am still getting the error on the nightly version of rust-analyzer, in my case in a call to `futures::try_join!`.

_Originally posted by @Victor-Savu in https://github.com/rust-analyzer/rust-analyzer/issues/6716#issuecomment-743399274_

**Comment by peku33** [comment]
Same here.
This affect all futures macros, like select, join etc.

**Comment by edwin0cheng** [comment]
Maybe related to #6811

**Comment by flodiebold** [comment]
@Victor-Savu @peku33 Please check this again with the next nightly or current master.

**Comment by Victor-Savu** [comment]
Unfortunately I can still reproduce the problem. Here is a minimal reproduction:

> Extension version: 0.3.418-nightly

`cargo.toml`:
```toml
[package]
name = "rust-analyzer-error"
version = "0.1.0"
edition = "2018"

[dependencies]
tokio = { version = "0.2", features = ["macros"] }
futures = "0.3"
```

`main.rs`:
```rust
#[tokio::main]
async fn main() {
    futures::join!(
        async {
            println!("Hello, world!");
        },
        async {
            println!("Hello, world!");
        }
    );
}
```

I used `join!` instead of `try_join!` to keep the example as short as possible but the error is the same:
> could not resolve macro `$crate::try_join_internal`

**Comment by flodiebold** [comment]
`join_internal` seems to be a `proc_macro_hack` macro; that's probably the problem.

**Comment by peku33** [comment]
So this is a problem with rust-analyzer or futures?
At the moment one of the most popular crates with 3M recent downloads (futures) triggers this error for basic usage.

**Comment by flodiebold** [comment]
Any error that rust-analyzer produces in code that compiles is a problem in rust-analyzer.

You can disable the diagnostic by adding `macro-error` to the `rust-analyzer.diagnostics.disabled` setting.

**Comment by lnicola** [comment]
Triage: this still happens even after #7145.

**Comment by lnicola** [comment]
Ah, it's a `proc_macro_hack` thing.

**Comment by bionicles** [comment]
still getting this, any easy way to disable it temporarily?

**Comment by Victor-Savu** [comment]
@bionicles if you are using VSCode, create a file `.vscode/settings.json` at the root of your project folder containing:

```json
{
    "rust-analyzer.diagnostics.disabled": [
        "macro-error"
    ]
}
```

**Comment by lolgeny** [comment]
It's annoying that with `futures::join`, with this, rust-analyser can't infer types for the output. Is there any hacky fix to get that working?

**Comment by mountainfloating** [comment]
> @bionicles if you are using VSCode, create a file `.vscode/settings.json` at the root of your project folder containing:
> 
> ```json
> {
>     "rust-analyzer.diagnostics.disabled": [
>         "macro-error"
>     ]
> }
> ```

Doesn't work for me with futures::select!. Any other way i can disable this?

**Comment by jonas-schievink** [comment]
The diagnostic is now called `unresolved-macro-call`

**Comment by Diggsey** [comment]
This also occurs with Yew's `html!` macro.

**Comment by KyGost** [comment]
This occurs with [fstrings](https://github.com/danielhenrymantilla/fstrings-rs)'s macros (seemably all?).
Disabling `unresolved-macro-call` provided a good temporary fix! Thanks.

**Comment by Ciantic** [comment]
If the problem is `proc_macro_hack`, then I think futures needs to be just updated? Maybe there we should create an issue in futures repo to remove the proc_macro_hack. After all the proc_macro hack repo says: 

> Proc_macro_hack Note: As of Rust 1.45 this crate is superseded by native support for #[proc_macro] in expression position. Only consider using this crate if you care about supporting compilers between 1.31 and 1.45.

**Comment by Ciantic** [comment]
It's now fixed in the futures package! It works now with the futures git repo.

![image](https://user-images.githubusercontent.com/64731/117128055-966f0780-ada5-11eb-952b-fca3f544b674.png)

I think this can be closed, though futures is not released yet, it will start to work in futures 0.4.

**Comment by jonas-schievink** [comment]
I'd rather keep it open. It will probably take years until the ecosystem uses futures 0.4

**Comment by taiki-e** [comment]
I will backport that patch to 0.3, so no need to wait for 0.4. (See the tag of the PR that fixed the issue)

**Comment by taiki-e** [comment]
Released futures 0.3.15 which includes the fix for this issue.

**Comment by regexident** [comment]
I recently started getting the same error with `log`'s macros, such as `log::debug!(…)` on v0.2.599, for what it's worth.

Should I open a separate issue or is this reasonably related?

**Comment by jonas-schievink** [comment]
Doesn't sound like the same issue

**Comment by sebaslopez01** [comment]
> I've just ran into this issue recently with [pyo3](https://github.com/PyO3/PyO3). you can try it yourself with pyo3 numpy example project: https://github.com/PyO3/rust-numpy/tree/main/examples/simple-extension.

Me too, I am also disabling it.

**Comment by Jason5Lee** [comment]
I got the problem in [ifmt](https://docs.rs/ifmt/0.3.3/src/ifmt/lib.rs.html).

**Comment by alexdevteam** [comment]
I've ran into this issue with `tokio::select!` and the `thiserror` crate `Error` derive.

**Comment by lnicola** [comment]
@alexdevteam `thiserror` used to work. Which compiler version are you using, and can you prepare a small test case?

**Comment by ModProg** [comment]
Got this error with `serde_with::flattened_maybe` as well, trying to resolve `$crate::serde::Deserialize`

**Comment by huntc** [comment]
In my particular case, I was getting the problem with Yew's macros. Other macros from other libraries like Embassy seemed to work.

To work-around my issue, my settings.json looks like:

```
{
    "rust-analyzer.diagnostics.disabled": [
        "unresolved-proc-macro"
    ]
}
```

**Comment by gmelodie** [comment]
> In my particular case, I was getting the problem with Yew's macros. Other macros from other libraries like Embassy seemed to work.
> 
> To work-around my issue, my settings.json looks like:
> 
> ```
> {
>     "rust-analyzer.diagnostics.disabled": [
>         "unresolved-proc-macro"
>     ]
> }
> ```

For anyone using CoC with neovim:
1. open nvim
2. `:CocConfig` (it'll open CoC's config file)
3. paste the above into this file
4. Save and reopen nvim

**Comment by Veykril** [comment]
I'll close this issue, a lot of comments in here are about very different issues so we gain more from people opening new specific issues instead.
